File: frontend/src/api/materials.ts
import { http } from './http'

export interface MaterialItem {
  id: string
  project_id: string
  original_name: string
  file_size: number
  mime_type: string
  parse_status: string
  parse_error: string | null
  created_at: string
}

export interface MaterialListResponse {
  items: MaterialItem[]
  total: number
}

export function getMaterials(projectId: string): Promise<MaterialListResponse> {
  return http.get(`/api/v1/projects/${projectId}/materials`)
}

export function uploadMaterial(projectId: string, file: File): Promise<MaterialItem> {
  const formData = new FormData()
  formData.append('file', file)
  return fetch(`/api/v1/projects/${projectId}/materials`, {
    method: 'POST',
    body: formData,
  }).then(async (res) => {
    if (!res.ok) {
      const text = await res.text().catch(() => '')
      throw new Error(`HTTP ${res.status}: ${text || res.statusText}`)
    }
    return res.json()
  })
}

export function deleteMaterial(projectId: string, materialId: string): Promise<void> {
  return http.delete(`/api/v1/projects/${projectId}/materials/${materialId}`)
}

File: backend/app/api/routes/materials.py
"""材料 CRUD API。"""

from fastapi import APIRouter, Depends, UploadFile, File, HTTPException
from sqlalchemy.orm import Session

from app.db.session import get_db
from app.models.material import Material
from app.models.project import Project
from app.schemas.material import MaterialResponse, MaterialListResponse
from app.services.file_storage import FileStorage

router = APIRouter(prefix="/projects/{project_id}/materials", tags=["materials"])


def _get_project(project_id: str, db: Session) -> Project:
    project = db.query(Project).filter(Project.id == project_id).first()
    if not project:
        raise HTTPException(status_code=404, detail="项目不存在")
    return project


@router.post("", response_model=MaterialResponse, status_code=201)
def upload_material(
    project_id: str,
    file: UploadFile = File(...),
    db: Session = Depends(get_db),
):
    _get_project(project_id, db)
    storage = FileStorage()
    content = file.file.read()
    try:
        stored_path = storage.save(project_id, file.filename or "unnamed", content)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))

    material = Material(
        project_id=project_id,
        original_name=file.filename or "unnamed",
        stored_path=str(stored_path),
        file_size=len(content),
        mime_type=file.content_type or "application/octet-stream",
    )
    db.add(material)
    db.commit()
    db.refresh(material)
    return material


@router.get("", response_model=MaterialListResponse)
def list_materials(project_id: str, db: Session = Depends(get_db)):
    _get_project(project_id, db)
    items = db.query(Material).filter(Material.project_id == project_id).order_by(Material.created_at.desc()).all()
    return MaterialListResponse(items=items, total=len(items))


@router.delete("/{material_id}", status_code=204)
def delete_material(project_id: str, material_id: str, db: Session = Depends(get_db)):
    _get_project(project_id, db)
    material = db.query(Material).filter(Material.id == material_id, Material.project_id == project_id).first()
    if not material:
        raise HTTPException(status_code=404, detail="材料不存在")
    import os
    if os.path.exists(material.stored_path):
        os.remove(material.stored_path)
    db.delete(material)
    db.commit()
